contrib/babelfishpg_tsql/src/table_variable_mvcc.c

```diff
@@ -0,0 +1,1094 @@
+	HTSV_Result res;
+
+	res = TVHeapTupleSatisfiesVacuumHorizon(htup, buffer, &dead_after);
+
+	if (res == HEAPTUPLE_RECENTLY_DEAD)
+	{
+		Assert(TransactionIdIsValid(dead_after));
+
+		if (GlobalVisTestIsRemovableXid(snapshot->vistest, dead_after))
+			res = HEAPTUPLE_DEAD;
+	}
+	else
+		Assert(!TransactionIdIsValid(dead_after));
+
+	return res != HEAPTUPLE_DEAD;
+}
+
+/*
+ * HeapTupleSatisfiesVisibility
+ *		True iff heap tuple satisfies a time qual.
+ *
+ * Notes:
+ *	Assumes heap tuple is valid, and buffer at least share locked.
+ *
+ *	Hint bits in the HeapTuple's t_infomask may be updated as a side effect;
+ *	if so, the indicated buffer is marked dirty.
+ */
+bool
+TVHeapTupleSatisfiesVisibility(HeapTuple tup, Snapshot snapshot, Buffer buffer)
+{
+    switch (snapshot->snapshot_type)
+    {
+        case SNAPSHOT_MVCC:
+            return TVHeapTupleSatisfiesMVCC(tup, snapshot, buffer);
+            break;
+        case SNAPSHOT_ANY:
+            return TVHeapTupleSatisfiesAny(tup, snapshot, buffer);
+            break;
+        case SNAPSHOT_SELF:
+            return TVHeapTupleSatisfiesSelf(tup, snapshot, buffer);
+            break;
+        case SNAPSHOT_DIRTY:
+            return TVHeapTupleSatisfiesDirty(tup, snapshot, buffer);
+            break;
+        case SNAPSHOT_TOAST:
+             return TVHeapTupleSatisfiesToast(tup, snapshot, buffer);
+             break;
+        case SNAPSHOT_NON_VACUUMABLE:
+             return TVHeapTupleSatisfiesNonVacuumable(tup, snapshot, buffer);
+             break;
+        case SNAPSHOT_HISTORIC_MVCC:
+        //     return HeapTupleSatisfiesHistoricMVCC(tup, snapshot, buffer);
+        //     break;
+
+
+            ereport(WARNING, (errmsg("[TableVariableAM] Unsupported snapshot type %d", snapshot->snapshot_type)));
+            break;
+    }
+
+    return false;				/* keep compiler quiet */
+
+}
+
+
+/*
+ * TVHeapTupleSatisfiesUpdate
+ *
+ * Counterpart of HeapTupleSatisfiesUpdate.
+ * The only difference is this function is not rollback sensitive.
+ */
+TM_Result
+TVHeapTupleSatisfiesUpdate(HeapTuple htup, CommandId curcid,
+                         Buffer buffer)
+{
+    HeapTupleHeader tuple = htup->t_data;
+
+    Assert(ItemPointerIsValid(&htup->t_self));
+    Assert(htup->t_tableOid != InvalidOid);
+
+    if (!HeapTupleHeaderXminCommitted(tuple))
+    {
+        if (HeapTupleHeaderXminInvalid(tuple))
+            return TM_Invisible;
+
+        else if (TransactionIdIsCurrentTransactionId(HeapTupleHeaderGetRawXmin(tuple)))
+        {
+            if (HeapTupleHeaderGetCmin(tuple) >= curcid)
+                return TM_Invisible;    /* inserted after scan started */
+
+            if (tuple->t_infomask & HEAP_XMAX_INVALID)  /* xid invalid */
+                return TM_Ok;
+
+            if (HEAP_XMAX_IS_LOCKED_ONLY(tuple->t_infomask))
+            {
+                TransactionId xmax;
+
+                xmax = HeapTupleHeaderGetRawXmax(tuple);
+
+                /*
```

Comment:
no parallel locking. But it could be locked by an aborted subtransaction at that point the transaction wont appear as it belongs to the current transaction and it's considered "another" transaction